Add Active Storage to Rails

rafaelfranca opened the pull request:
Need to address a few issues:

- [ ] How should the initial migration be seeded? I'm thinking probably just generate it in db/migrations on new apps.
- [x] Move the `direct_upload: true` convenience option from the activestorage helper into actionview.
- [x] Address how Rails' CI can run the cloud service integration tests.

jeremy committed:
Disk storage: ensure URLs end with the blob filename since some user agents don't respect Content-Disposition filename

Commit:
Merge pull request #22 from jeremy/disk-filename-url

Disk storage: ensure URLs end with the blob filename

Commit:
Merge pull request #21 from jeremy/disk-binary-io

Disk service: use binary IO throughout, not UTF-8

jeremy committed:
Mirror: explicit primary service and list of mirrors

Pass separate primary service and list of mirrors rather than treating
the first of the services list as the primary. Nice fit for keyword args,
and something we've long wanted in the equivalent Basecamp file repository.

Upload returns the results of the underlying service uploads rather than
the io.rewind result. Rewind before uploading rather than afterward, and
demonstrate that behavior with a test.

Test that more than one mirror works.

jeremy committed:
Configure services that reference other services

* Move service configuration from the Engine to Service
* Delegate configuration mechanics to internal Service::Configurator
* Delegate service building to the concrete Service classes, allowing
  them to configure composed services.
* Implement for the Mirror service.

jeremy committed:
Travis CI 💡

jeremy committed:
Service: clarify Service.build arguments

First arg is config for the service we're instantiating.

Second arg is service configurations so we can look up and configure
other services by name.

jeremy committed:
Clarify how a service can build other composed services

* Service.build takes the literal YAML config hash for the service and a
  reference to the Configurator that's doing the building.
* Services that compose additional services can use the Configurator to
  look them up and build them by name. See MirrorService for an example.

References #23

Commit:
Match new configurator needs

Commit:
Merge pull request #26 from jeremy/service-configurator

Clarify how a service can build other composed services

Commit:
Use class methods scope now that we have multiple

dixpac committed:
Symbolize all keys inside configuration nested hash

Since configuration is a nested hash we need to symbolize all keys
of the hash. Othervise fetcing will fail on start

Commit:
Ensure binary encoding for downloading

Commit:
Example of keyfile specification

Commit:
Merge pull request #27 from dixpac/fix_configuration_keys_symbolization

Symbolize all keys inside configuration nested hash

Commit:
Instrument and log the services

Commit:
Test URL generation for S3 and Disk

Commit:
Proper logging is now in place

Commit:
Copypasta comments

# Conflicts:
#	lib/active_storage/engine.rb
#	lib/active_storage/service.rb
#	lib/active_storage/service/disk_service.rb
#	lib/active_storage/service/s3_service.rb
#	test/service/s3_service_test.rb
#	test/test_helper.rb

Commit:
Direct uploads for S3

Commit:
Accept sgids for existing blobs created via direct upload as an attachable

dixpac committed:
Add basic tests to the Configurator#build (#28)

Commit:
Explore regular polymorphic associations rather than record_gid

jeremy committed:
Configurator tests: work against test-local config

So tests pass when service configs aren't set up.

References #28

jeremy committed:
Tests: Dir.mktmpdir neatly wraps up tmpdir + join

dixpac committed:
Scope aws-skd to version 2 (#34)

Since we use new aws-sdk API, I've scoped aws-sdk version
so someone doesn't accidentaly installs wrong version during the
development.

georgeclaghorn committed:
Expose chunked downloads

georgeclaghorn committed:
Depend on Rails > 5.1

georgeclaghorn committed:
Revert "Depend on Rails > 5.1"

This reverts commit 53d5384ac265ead44b25eec5f8f5020568184da2.

Commit:
Very incomplete first stab

jeremy committed:
S3: slim down service implementation (#40)

* Use simple core API for duck-type compat with other clients
* initialize: accept an existing client
* initialize: accept arbitrary client args instead of a fixed, required set
* download: use native get_object streaming, no need to implement range requests
* exists?: use head_object (which returns immediately) rather than waiting for existence

Commit:
Revert "S3: slim down service implementation (#40)" (#41)

This reverts commit 6d3962461fb8d35fc9538d685fee96267663acf2.

quartzmo committed:
Update GCSService#url

Update google-cloud-storage dependency to 1.3
Refactor arguments to Google::Cloud::Storage::File#signed_url

georgeclaghorn committed:
Merge pull request #42 from quartzmo/gcs-site-url

Update GCSService#url

dixpac committed:
Remove few ivars from gcs_service implementation

georgeclaghorn committed:
Merge pull request #43 from dixpac/small_gcs_service_refactoring

Remove unnecessary variables from gcs_service implementation

dixpac committed:
Fix regular expression on s3 test URL generation test (#44)

So tests are passing if the bucket name is rails-active storage.
But developers specify their own s3 tests configuration (in my case was
activestorage-test) then this regex fails.

Also the first part is dynamic and based on bucket name and region

cristianbica committed:
Added rubocop / codeclimate config and fixed current offenses (#45)

seanhandley committed:
Allow custom endpoints for S3.

dixpac committed:
Revert back to the original implementaion with varaibles

Revert `exist? and url` to the original implementation.
Since the new one doesn't provide any benefits and makes implementation
harder to follow.

georgeclaghorn committed:
Merge pull request #51 from dixpac/restore_intermediate_variables

Revert back to the original implementaion with varaibles

Commit:
Don't fail on boot because of missing S3 keys when S3 is not used

Commit:
Fail early if no storage service is specified

georgeclaghorn committed:
Merge pull request #53 from colorfulfool/error-messages

Useful error messages for two possible misconfigurations

georgeclaghorn committed:
Depend on Rails >= 5.2.0.alpha

SGospodinov committed:
Fixing logger to work with Rails 5.2.0.alpha

georgeclaghorn committed:
Merge pull request #55 from SGospodinov/logger-for-5.2.0

Fixing logger to work with Rails 5.2.0.alpha

georgeclaghorn committed:
Put conditional back

georgeclaghorn committed:
Merge pull request #48 from seanhandley/hosted_s3

Allow custom endpoints for S3.

georgeclaghorn committed:
Swap branches

Commit:
Fix a typo in S3Service

georgeclaghorn committed:
Merge pull request #56 from colorfulfool/fix-s3-service

Fix a typo in S3Service

georgeclaghorn committed:
Fix S3 direct upload test

georgeclaghorn committed:
Splat options

michaelherold committed:
Add direct upload support to GCS service

georgeclaghorn committed:
Remove unused require

jtperreault committed:
Replace hard-coded S3 bucket name with configured bucket

georgeclaghorn committed:
Use descriptive new freeze_time helper

baerjam committed:
Accept S3 upload options (e.g. server_side_encryption)

Commit:
Merge branch 'master' into variants

Commit:
Double confetti

Commit:
Easier access to the variant of a blob

Commit:
Follow AR like naming of factory method

Commit:
Do real transformations in a safe way

Commit:
Use a unique blob variant key for storage

Commit:
Use the direct accessor

Commit:
Extract routes.rb to engine location for auto configuration

Commit:
Move controllers to default engine location for auto loading

Commit:
Fix-up variants controller

Commit:
Make processing an explicit step

Commit:
Consistent naming

Commit:
True is the same as no arguments

dixpac committed:
Add instruction to install the gem from the master (#65)

Commit:
Better naming

Commit:
Accept that this is a full-Rails engine

Commit:
Mount routes on the engine

Commit:
Hacky way to mount routes for engine controller tests

Commit:
Extract variation value object

Commit:
Precise naming

Commit:
Over-indented

Commit:
Fix and test VariantsController

Commit:
Proper require

Commit:
Provide directed URL as well as resolving

Commit:
Actually we just want them mounted straight

Commit:
Fix parens after inline block

Commit:
Extract test helper for image blob fixtures

Commit:
Anemic intro

Commit:
Test actual transformations

Commit:
Test actual transformation via controller too

Commit:
Quick example of variants

Commit:
Merge branch 'master' into variants

Commit:
Mention need for mini_magick with variants

Commit:
Merge pull request #63 from rails/variants

On-demand variants

koic committed:
Fix RuboCop offenses and warnings

dixpac committed:
Move all controller tests to controller/ dir

Commit:
Merge pull request #67 from dixpac/move_controller_tests

Move all controller tests to controllers dir

dixpac committed:
Assume Rails is defined

Commit:
Don't need to validate transformations actually

Since they're only ever generated in signed form. Users never have direct access to dictate transformations.

Commit:
Move storage_services.yml to config for consistency

Commit:
Unused, we can extract from it out-of-repo

Commit:
Move models and jobs to the app setup

Follow engine conventions more closely

Commit:
Move model tests to models directory

Commit:
Fix extension to run test automatically

Commit:
No need for explicit requires any more

Commit:
Nix more needless requires

Commit:
Even more needless requires

Commit:
Disk controller must rely on key alone

Otherwise it can't be used to display variants. It's better anyway since all other services won't know about blobs either. Better simulation. Closes #71

Commit:
Switch to a single message verifier

No need for this proliferation

Commit:
Switch to simpler signed_id for blob rather than full GlobalID

We don't need to lookup multiple different classes, so no need to use a globalid.

Commit:
Provide a BlobsController for stable blob URLs

We need to have stable urls for blobs and variants or caching won't work. So provide a controller that can give that and redirect to the service URL upon lookup.

Commit:
VerifiedKeyWithExpiration no longer needed

Thanks to rails/rails#29854! This does mean that we now depend on rails/rails master.

Commit:
Update the README with more explicit expectation setting

kaspth committed:
Use parsed_body to auto parse the response as JSON.

kaspth committed:
Collapse indeces per Jeremy.

kaspth committed:
Fix attaching with standard Rails associations.

Removes needless ivar caching (a Rails association handles that).

Inserts a reload and a nil assign, since the association proxy doesn't
seem to that it's been destroyed through `purge`.

Commit:
Specify verification purposes

Commit:
Merge branch 'master' of github.com:rails/activestorage

Commit:
Merge branch 'master' into explore-polymorphism

georgeclaghorn committed:
Require mini_magick when it's used

Commit:
Work-around until @response.parsed_body problem is solved

Commit:
Work-around until @response.parsed_body problem is solved

Commit:
Return to same level of abstraction

Commit:
Return to same level of abstraction

Commit:
Merge branch 'master' into explore-polymorphism

Commit:
Add and test preloading scope

Commit:
No more GlobalID

Commit:
Demonstrate preloading in example

Commit:
Merge pull request #32 from rails/explore-polymorphism

Use regular polymorphic associations rather than record_gid

Commit:
Fix attaching

Can use the associations now

Commit:
Schema out of date and now obvious

Commit:
Still need GlobalID for PurgeJob serialization

Commit:
Tell service which content-type to use for the response

georgeclaghorn committed:
Don't annotate template class with :nodoc:

georgeclaghorn committed:
Temporarily skip variant tests pending a fix

georgeclaghorn committed:
Skip controller test, too

Commit:
Refer to the yielded app

Commit:
Merge branch 'master' of github.com:rails/activestorage

Commit:
Everything under app/ is eager loaded, don't want that for service

Since it references all the specific cloud services that are intended only to be loaded on demand.

Commit:
Blob/Variant#url -> #service_url to emphasize this URL isn't to be public

georgeclaghorn committed:
Verify direct upload checksums

Closes #74.

Commit:
Basic documentation for all the models

Commit:
Start on docs for lib

Commit:
Merge branch 'master' of github.com:rails/activestorage

Commit:
Finish basic documentation for controllers

Commit:
Basic documentation for job

Commit:
Follow the same copyright format as the other Rails frameworks

Commit:
Add standard version setup

Commit:
Models are autoloaded per engine standards

Commit:
Fix blob associations

cc @javan

Commit:
Trivial typo fix

Only one character, but it needs to be done

Commit:
Document the rest of lib

Commit:
Flesh out the README a bit more

Commit:
Did that

georgeclaghorn committed:
Deep merge

georgeclaghorn committed:
Add direct upload support to the disk service

colorfulfool committed:
Compare images by selected attributes (size, colorspace)

georgeclaghorn committed:
Avoid creating unnecessary blobs

georgeclaghorn committed:
Check variant image properties directly

georgeclaghorn committed:
Fix test failure in CI due to platform differences

Commit:
Fix broken links in README

dixpac committed:
Introduce the Dummy. (#70)

georgeclaghorn committed:
Validate Content-Length

georgeclaghorn committed:
Use blob service URL methods

georgeclaghorn committed:
Separate show and direct upload tests

javan committed:
Add JavaScript direct upload support (#81)

javan committed:
Add JavaScript linting (#82)

javan committed:
Minor constructor signature tweaks

javan committed:
Add npm package details

Commit:
Add a direct_upload: true option to file_field_tag and Form#file_field

Then users don't have to bother with manually referring to internal routes.

Commit:
Add Blob#type as a StringInquirer

Commit:
Must always return the options even if they werent converted

Commit:
Convert type inquiry into root predicates for base types alone instead

Commit:
Include default storage configuration

javan committed:
Configure per-service request headers for direct uploads (#83)

* Configure per-service request headers for direct uploads

* Fix header hashes

javan committed:
activestorage.js 0.1.1

dixpac committed:
Depend on Ruby >=2.2.2 (#85)

dixpac committed:
Azure Storage support (#36)

* Microsoft Azure storage support

* Add support for Microsoft Azure Storage

* Comply with the new headers implementation

rafaelfranca committed:
Add 'activestorage/' from commit '3f4a7218a4a4923a0e7ce1b2eb0d2888ce30da58'

git-subtree-dir: activestorage
git-subtree-mainline: 0d58e7e478e79c2d6b2a39a4444d2a17a903b2a6
git-subtree-split: 3f4a7218a4a4923a0e7ce1b2eb0d2888ce30da58

Commit:
Use standard Rails layout for gemspec

Commit:
Initialize changelog

Commit:
Bring activestorage dependencies into the general Gemfile

Commit:
Require activestorage per default with asset pipeline

Commit:
No longer used

Commit:
Resolve gemspec typo

Commit:
Use config/storage.yml as part of the skeleton and loading

Commit:
Load activestorage by default

Commit:
Follow form from actioncable

Commit:
Fix gitignore to be relative

Commit:
Create the skeleton directories needed and ignore accordingly

Commit:
Directory and config installation now done by app_generator

Commit:
Move to new storage.yml location

Commit:
Move the direct_upload: true convenience option from the activestorage helper into actionview

Commit:
Setup travis to be able to run CI tests against S3

georgeclaghorn commented on activestorage/.rubocop.yml:
```diff
@@ -0,0 +1,125 @@
+
+# Use my_method(my_arg) not my_method( my_arg ) or my_method my_arg.
+Lint/RequireParentheses:
+  Enabled: true
```
I don’t think we need to import this. The top-level RuboCop config is sufficient.

georgeclaghorn commented on activestorage/.travis.yml:
```diff
@@ -0,0 +1,25 @@
+
+script:
+  - bundle exec rake
+  - yarn lint
```
We don’t need to import this, either.

georgeclaghorn commented on railties/lib/rails/generators/rails/app/templates/config/storage_services.yml:
```diff
@@ -0,0 +1,27 @@
+# mirror:
+#   service: Mirror
+#   primary: local
+#   secondaries: [ amazon, google ]
```
Should this template be renamed to match the name of the file it’s used to generate?

georgeclaghorn commented on actionview/lib/action_view/helpers/form_helper.rb:
```diff
@@ -2316,6 +2316,10 @@ def convert_to_legacy_options(options)
         end
+
+        def convert_direct_upload_option_to_url(options)
+          options.merge('data-direct-upload-url': options.delete(:direct_upload) ? rails_direct_uploads_url : nil).compact
```
This file needs to require `"active_support/core_ext/hash/compact"` now.

lugray commented on activestorage/README.md:
```diff
@@ -0,0 +1,151 @@
+| `direct-upload:end` | `<input>` | `{id, file}` | A direct upload has ended. |
+| `direct-uploads:end` | `<form>` | None | All direct uploads have ended. |
+
+## Compatibility & Expectations
```
This section needs to be re-written for living inside rails.

claudiob commented on activestorage/test/service/configurations.yml:
```diff
@@ -0,0 +1,30 @@
+  access_key_id: <%= ENV["AWS_ACCESS_KEY_ID"] %>
+  secret_access_key: <%= ENV["AWS_SECRET_KEY"] %>
+  region: us-east-2
+  bucket: rails-ci-activestorage
```
Because of these lines, tests are run by default against S3 but not against GCS and Azure.

However, a rails developer might not have access to an S3 instance to make these tests pass.

How do you feel about having something on the line of Active Record's tests, where you can run:

```
bundle exec rake test:s3
bundle exec rake test:gcs
bundle exec rake test:azure
```

if you explicitly want to test those components, whereas as default you only test disk storage?

dhh replied:
I like that upgrade 👍

claudiob committed:
Let ActiveStorage tests pass only for Disk

If you have a "service/configurations.yml" file, but every single line is
commented out, then an error occurs when running tests:

```
git:active-storage-import~/code/rails/activestorage$ rake
~/code/rails/activestorage/test/test_helper.rb:17:in `<top (required)>': undefined method `deep_symbolize_keys' for false:FalseClass (NoMethodError)
	from ~/code/rails/activestorage/test/controllers/direct_uploads_controller_test.rb:1:in `require'
```

The reason is that `YAML.load(..an empty file content..)` simply returns `false`, and not `{}`.

This PR fixes this behavior so tests can also run when no remote service
is available.

claudiob commented on activestorage/test/test_helper.rb:
```diff
@@ -0,0 +1,54 @@
+
+require "yaml"
+SERVICE_CONFIGURATIONS = begin
+  YAML.load(ERB.new(Pathname.new(File.expand_path("../service/configurations.yml", __FILE__)).read).result).deep_symbolize_keys
```
If you have a "service/configurations.yml" file, but every single line is commented out, then an error occurs when running tests:

```
git:active-storage-import~/code/rails/activestorage$ rake
~/code/rails/activestorage/test/test_helper.rb:17:in `<top (required)>': undefined method `deep_symbolize_keys' for false:FalseClass (NoMethodError)
	from ~/code/rails/activestorage/test/controllers/direct_uploads_controller_test.rb:1:in `require'
```

The reason is that `YAML.load(..an empty file content..)` simply returns `false`, and not `{}`.

If you think this should be fixed, you can merge this:
https://github.com/rails/rails/compare/active-storage-import...claudiob:only-disk?expand=1

I haven't explicitly opened a PR because it's a commit against a branch different than `master`, so I wasn't sure if that was the right approach.

claudiob commented on activestorage/test/controllers/direct_uploads_controller_test.rb:
```diff
@@ -0,0 +1,121 @@
+        assert_equal 6, details["byte_size"]
+        assert_equal checksum, details["checksum"]
+        assert_equal "text/plain", details["content_type"]
+        assert_match /#{SERVICE_CONFIGURATIONS[:s3][:bucket]}\.s3.(\S+)?amazonaws\.com/, details["direct_upload"]["url"]
```
Should the `.` after `s3` be `\.`? 
That is, it should match an actual dot and not any character.

claudiob replied:
This test is currently failing if the the name of the AWS bucket includes a `.`.

For instance, if the bucket is called `rails.test` then the direct upload URL returned by AWS can be:

"https://s3.amazonaws.com/rails.test/xucyA48o9DbySmKBP…"

rather than

"https://rails.test.s3.amazonaws.com/xucyA48o9DbySmKBP…"

guilleiguaran replied:
👍

claudiob replied:
Fided in #30059

claudiob replied:
Fixed in #30059

claudiob commented on activestorage/test/service/s3_service_test.rb:
```diff
@@ -0,0 +1,52 @@
+    end
+
+    test "signed URL generation" do
+      assert_match /#{SERVICE_CONFIGURATIONS[:s3][:bucket]}\.s3.(\S+)?amazonaws.com.*response-content-disposition=inline.*avatar\.png.*response-content-type=image%2Fpng/,
```
See my comment above https://github.com/rails/rails/pull/30020#discussion_r130663525

This test fails in the same way.

claudiob replied:
Fixed in #30059

claudiob commented on Gemfile:
```diff
@@ -92,6 +92,19 @@ group :cable do
 end
 
+group :storage do
+  gem "httparty"
```
Looks like "httparty" is only used in two tests to make PUT requests to S3 and GCS.

How do you feel about dropping this dependency and simply using `Net::HTTP::Put` in those two tests?

I could make that change if you agree. Thanks.

claudiob replied:
Opened a PR: https://github.com/rails/rails/pull/30032

guilleiguaran commented:
@claudiob feel free to open a PR against this branch with the proposed changes!!!

claudiob commented:
@guilleiguaran Thanks! I wasn't sure about that… I will open them now starting with https://github.com/rails/rails/pull/30030

claudiob committed:
Don't depend on HTTParty

"httparty" is only added in #30020 to write two tests to make PUT requests
against S3 and GCS.

The same requests can be made with net/http, removing a dependency from the Gemfile.

guilleiguaran committed:
Merge pull request #30032 from claudiob/remove-httparty

Remove HTTParty dependency

guilleiguaran committed:
Merge pull request #30030 from claudiob/only-disk

Let ActiveStorage tests pass only for Disk

Commit:
Add Azure configuration to Active Storage template

With Active Storage supporting Azure, we should show people how to set it up.

Incorporates the minor changes originally in https://github.com/rails/activestorage/pull/88 as well

guilleiguaran committed:
Merge pull request #30034 from ffmike/patch-1

Add Azure configuration to Active Storage template

Commit:
Rely on top-level files

Commit:
We moved to storage.yml

Commit:
Add require

Commit:
Drop README terms no longer relevant

Commit:
Cut down on installation advice

Commit:
Follow same format as action_cable

Commit:
Add the storage.yml file by default

Commit:
Prevent interpolation of ERB elements

Commit:
Merge branch 'master' into active-storage-import

Commit:
Remove circular dependency

Commit:
Fix dummy app for inclusion in Rails

Commit:
Skip if credentials aren't provided

ytbryan committed:
Add Azure to readme 

previously https://github.com/rails/activestorage/pull/86

kamipo committed:
Merge pull request #30035 from ytbryan/patch-1

Add Azure to readme [ci skip]

Commit:
Add Azure to the available services list

kamipo committed:
Merge pull request #30040 from rbarysas/active-storage-import

Add Azure to the available services list [ci skip]

dixpac committed:
Depend on offical azure-core

No need to depend on my fork anymore, they've just release fix
officially.

Commit:
Merge branch 'master' into active-storage-import

kamipo committed:
Fix `Style/StringLiterals` violations for Active Storage

```
% be rubocop -a --only Style/StringLiterals activestorage
Inspecting 74 files
........................................CCCCCCCCCC.C........CC.......C.C..

(snip)

74 files inspected, 31 offenses detected, 31 offenses corrected
```

rafaelfranca committed:
Merge pull request #30046 from kamipo/fix_string_literals_violations_for_active_storage

Fix `Style/StringLiterals` violations for Active Storage

Commit:
Add Azure gems to README

Commit:
Update README.md

Commit:
Use release azure-core gem

Commit:
Tweak requires

Omit azure-core (it's required by azure-storage), remove require:false to match other points.

schneems committed:
Merge pull request #30053 from ffmike/patch-2

Add Azure gems to README

Commit:
Can't initialize a message verifier without a key base

Commit:
Generating the app is where the exception is now raised

claudiob committed:
Make Rubocop happier about ActiveStorage

Running `rubocop activestorage` before this commit resulted in 20 offenses.
This commit only fixes:

- Trailing whitespace detected
- Space inside } missing
- Put one space between the method name and the first argument.

The other offenses are left since they are intentional according to
@georgeclaghorn (https://github.com/rails/rails/pull/30061#issuecomment-319999190)

rafaelfranca committed:
There is no reason to single line methods here

I know those methods are unlikely to change but having one line method
is hard to read and also hard to modify.

claudiob commented on activestorage/lib/active_storage/migration.rb:
```diff
@@ -0,0 +1,27 @@
+class ActiveStorageCreateTables < ActiveRecord::Migration[5.1]
```
Do you think this class should be `nodoc` or needs to be documented?
Currently it's showing up when running `rake rdoc` at the very root (since it's not namespaced):

<img width="313" alt="list" src="https://user-images.githubusercontent.com/10076/28939973-eab68490-7847-11e7-9566-221874b22edf.png">

georgeclaghorn replied:
Neither, since it’s a template that’s copied into apps.

georgeclaghorn replied:
It should probably have an extension that communicates it’s a template and excludes it from documentation, or we should otherwise exempt it if that’s at all possible.

claudiob commented on activestorage/app/controllers/active_storage/blobs_controller.rb:
```diff
@@ -0,0 +1,22 @@
+# Take a signed permanent reference for a blob and turn it into an expiring service URL for download.
```
How are these four controllers in app/controllers/active_storage  intended to be used in a Rails app?
Are they meant as superclasses to inherit from?

georgeclaghorn replied:
> Are they meant as superclasses to inherit from?

No, though you could probably extend them if you so desired. Apps use them by sending requests to them, either via routes or `DiskService#url{,_for_direct_upload}`.

kaspth replied:
Active Storage is a Rails::Engine so the controllers (and their routes) are available in users' Rails apps 😄

Commit:
Active Storage loads AR earlier to extend it so require env later

Commit:
Including new default classes in loading test

Commit:
Same issue from AR getting loaded earlier

cc @rafaelfranca

rafaelfranca committed:
Make sure Action View  doesn't break with Active Storage

When Active Storage is not loaded and direct_upload is used on
file_field_tag we should not raise an exception.

rafaelfranca committed:
Refactor convert_direct_upload_option_to_url

Also make sure file_field doesn't mutate the original options passed in.

rafaelfranca committed:
Remove duplicated convert_direct_upload_option_to_url

FormHelper includes FormTagHelper so we don't need to define two methods

rafaelfranca committed:
Do not eager load ActiveRecord::Base

Everything inside the app directory of a engine is autoload/eager loaded automatically so we don't need to require them.

Commit:
Active Storage routes are now part of the default routes

It's worth considering whether we should hide these by default, but I'm kinda thinking no. It's very reasonable that someone would want to call these directly, so they should be documented.

Commit:
Merge branch 'active-storage-import' of github.com:rails/rails into active-storage-import

rafaelfranca committed:
We need to require the top file before individual one

rafaelfranca committed:
Merge pull request #30061 from claudiob/ast-rubocop

Make Rubocop happy about ActiveStorage

rafaelfranca committed:
Fix all rubocop violations

rafaelfranca committed:
Merge pull request #30041 from dixpac/change_azure_core_dependecy

Change azure core dependency

rafaelfranca committed:
Do not change the entire codebase style only because of active_storage

We are already removing the braces around hash parameters in the last
argument in other places so we should not change the entire codebase
because of two places.

Commit:
Deterministic comparisons please

Commit:
Convert to strings so array can be sorted deterministically

Commit:
Include active_storage in all generated application.rb files

And deal with a temporary test fix until we allow you to skip active storage.

rafaelfranca committed:
Do not call isolate_namespace on the application

What we want to test is that two different calls to isolate_namespace
with the same module doesn't change the original railtie. We can do that
defining two different railties.

We can't call in the application because this method is not supposed to
be called in an Application class.

rafaelfranca committed:
Make sure the test fails if the code chages

The test was passing before because it was not being testes correctly.

Now we create a different engine that is loaded before the already
exising and we make sure that the first call for isolate_namespace is
what takes effect.

claudiob committed:
Add Active Storage to README and release (#30065)

Before we forget...

claudiob committed:
Fix tests for AWS buckets that include a . (#30059)

If an AWS bucket name includes a `.` (e.g. `bucket.name`), then the canonical
URL for an object will start with "https://s3.amazonaws.com/bucket.name/"
and not with "https://bucket.name.s3.amazonaws.com/".

The URL tests have now been separated into two separate asserts, to ensure
that both the "s3.amazonaws.com" and the "bucket.name" components are included,
but not specifically in that order.

dixpac committed:
Rename Azure to AzureStorage (#30057)

kaspth commented:
Wups! It works! ❤️
